pypesto/hierarchical/optimal_scaling_approach/optimal_scaling_problem.py

```diff
@@ -0,0 +1,530 @@
+from typing import Dict, List, Tuple
+
+import numpy as np
+import pandas as pd
+
+from ...C import (
+    INNER_PARAMETER_BOUNDS,
+    LIN,
+    MEASUREMENT_CATEGORY,
+    MEASUREMENT_GROUP,
+    MEASUREMENT_TYPE,
+    ORDINAL,
+    REDUCED,
+    STANDARD,
+    TIME,
+    InnerParameterType,
+)
+from ..problem import (
+    InnerProblem,
+    _get_timepoints_with_replicates,
+    ix_matrices_from_arrays,
+)
+from .optimal_scaling_parameter import OptimalScalingParameter
+
+try:
+    import amici
+    import petab
+    from petab.C import OBSERVABLE_ID
+except ImportError:
+    pass
+
+
+class OptimalScalingProblem(InnerProblem):
+    """Class of the Optimal Scaling inner subproblem."""
+
+    def __init__(
+        self,
+        xs: List[OptimalScalingParameter],
+        data: List[np.ndarray],
+        method: str,
+    ):
+        """Construct the Optimal Scaling inner subproblem.
+
+        Parameters
+        ----------
+            xs:
+                List of `OptimalScalingParameter`s of the subproblem.
+            data:
+                The data of the problem.
+            method:
+                A string representing the method of the Optimal Scaling approach, either 'reduced' or 'standard'.
+        """
+        super().__init__(xs, data)
+        self.groups = {}
+        self.method = method
+
+        for group in self.get_groups_for_xs(InnerParameterType.OPTIMALSCALING):
+            self.groups[group] = {}
+            xs = self.get_xs_for_group(group)
+            self.groups[group]['num_categories'] = len(
+                {x.category for x in xs}
+            )
+            self.groups[group]['num_datapoints'] = np.sum(
```

Comment:
for list inputs, python `sum()` should be more efficient than first converting to `np.ndarray` for `np.sum()`